EventPhases/explainer.md

```diff
@@ -0,0 +1,501 @@
+
+ 1. Browser processing (style, layout, etc.)
+ 1. JavaScript event dispatch
+    1. Read phase
+       1. Task execution
+       1. Microtask queues drained
+    1. Read/Write
+       1. Task execution
+       1. Microtask queues drained
+    1. Write
+       1. Task execution
+       1. Microtask queues drained
+ 1. GOTO: 1
+
+In each turn, the `read` phase happens strictly before the `write` phase. We expect that style readback in the `write` phase will be allowed, but simply always return dirty information.
+
+For example:
+
+```js
+let foo = document.getElementById("foo");
+foo.addEventListener(
+  // Event type
+  "click", 
+  // Handler
+  (evt) => {
+    // Synchronous style readbacks are fine here
+    let w = foo.innerWidth;
+    let h = foo.innerHeight;
+    console.log(w, h);
+
+    // Writes are not allowed, and instead must 
+    // be scheduled. This could be written many 
+    // ways, but we imagine the Scheduler API 
+    // might be a good place for it:
+    scheduler.postTask(
+      () => {
+        let styles = foo.styleAttributeMap;
+        styles.set("width",  CSS.px(w + 10));
+        styles.set("height", CSS.px(h + 10));
+        // ...
+      },
+      { phase: "write" }
+    );
+  },
+  // Options
+  { phase: "read" }
+);
+
+// And using the new `when()` observable style:
+foo.when(
+  "click", 
+  (evt) => { /* ... */}, 
+  { phase: "read" }
+);
+```
+
+Syntax of the scheduling integration is TBD (e.g., should there be explicit `scheduler.postForRead()` and `scheduler.postForWrite()` API instead?). We could also imagine the event carrying scheduling methods for brevity:
+
+```js
+foo.when("click", 
+  (evt) => {
+    // ...
+    evt.on("write", () => { /* ... */ });
+  },
+  { phase: "read" }
+);
+```
+
+It's also unclear if any (or all) of these methods should return promises. See Considered Alternatives for variations that can.
+
+Extensions will also be required for the platform's existing scheduling APIs to allow them to "call" their phase when not triggering today's conjoined `read/write` phase. Examples include:
+
+- `setTimeout()`
+- `setInterval()`
+- `requestAnimationFrame()`
+- `requestIdleCallback()`
+- `postMessage()`
+- `queueMicrotask()`
+
+### Opt-in Mechanism
+
+Acheiving widespread adoption of these mechanisms is a daunting but important task, as any stray third party library can effectively destroy the performance of web applications today. As a result, we place a premium on alternatives that allow for incremental rewrites of existing code, as well as the ability to polyfill with minimal intrusion.
+
+The meat of our proposal (above) does not require any explicit page-level opt-in, but we can imagine enforcement modes that can set a global policy to strictly require use of them; e.g. via [Document Policy](https://github.com/WICG/document-policy/blob/main/document-policy-explainer.md):
+
+```html
+<html>
+  <head>
+    <meta http-equiv="Document-Policy" 
+          content="event-phases: required">
```

Comment:
This section got me thinking about browser extensions. Might a web author expect to protect themselves from extensions that can cause layout thrashing with a policy like this? Would such an expectation be reasonable under priority of constituencies?

Reply:
Good question. I'll add it to the Open Questions section that Alison prompted me to add.